### PR TITLE
client: hold reference for returned inode

### PR DESCRIPTION
**Pull request by ukernel**
CLient::make_request() returns pointer of target inode, but it does not
increase reference count of the returned inode. The inode may get freed
when CLient::make_request() releases MetaRequest.

The fix is hold a reference for the returned inode. Since there are many
of places use CLient::make_request() directly or indirectly, it's easy
to leak reference. This patch introduces a 'smart pointer' class and use
it to track the reference.

Signed-off-by: Yan, Zheng <zyan@redhat.com>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -1366,9 +1366,48 @@ void Client::dump_mds_requests(Formatter *f)
...
+    C_InodeRef tmp(client, in);
+    swap(tmp);
+    return *this;
+  }
+  C_InodeRef& operator=(const C_InodeRef& o) {
+    C_InodeRef tmp(o);
+    swap(tmp);
+    return *this;
+  }
+  Inode *get() const { return in; }
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Why did you want to explicitly invoke a get() function instead of having an implicit conversion from C_InodeRef to Inode*?

File: src/client/Client.cc [hunk 2]
```diff
@@ -9714,25 +9741,25 @@ int Client::_link(Inode *in, Inode *dir, const char *newname, int uid, int gid,
   return res;
 }
 
-int Client::ll_link(Inode *parent, Inode *newparent, const char *newname,
+int Client::ll_link(Inode *in, Inode *newparent, const char *newname,
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Why the name change, and did it need to go in the giant patch instead of in a separate one for easier review?

File: src/client/Client.cc [hunk 3]
```diff
@@ -1435,6 +1471,10 @@ int Client::verify_reply_trace(int r,
       }
     }
   }
+  if (r >= 0) {
+    assert(target.get());
+    ptarget->swap(target);
+  }
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Why this dance instead of using ptarget directly, as we did previously?
We could still clear it out if there's an error condition.

(I'm just curious about the things that inflate the diff line count; that doesn't mean they're a bad idea but they're harder to digest.)

File: src/client/Client.cc [hunk 4]
```diff
@@ -5209,32 +5249,32 @@ int Client::get_or_create(Inode *dir, const char* name,
   return 0;
 }
 
-int Client::path_walk(const filepath& origpath, Inode **final, bool followsym)
+int Client::path_walk(const filepath& origpath, C_InodeRef *end, bool followsym)
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Why the name change?

File: src/client/Client.cc [hunk 5]
```diff
@@ -8376,42 +8407,41 @@ int Client::ll_lookup(Inode *parent, const char *name, struct stat *attr,
...
-    }
+  rc = path_walk(fp, &in, false);
+  if (rc < 0) {
+    attr->st_ino = 0;
+    *out = NULL;
+    return rc;
+  } else {
+    assert(in.get());
+    fill_stat(in.get(), attr);
+    _ll_get(in.get());
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: I'm having trouble tracking down why we need to add an explicit lock access here. I presume we were previously grabbing one explicitly in the depths that is now handled by our use of C_InodeRef, but I can't find it.

File: src/client/Client.cc [hunk 6]
```diff
@@ -8555,12 +8585,10 @@ int Client::ll_setattr(Inode *in, struct stat *attr, int mask, int uid,
   tout(cct) << attr->st_atime << std::endl;
   tout(cct) << mask << std::endl;
 
-  Inode *target = in;
-  int res = _setattr(in, attr, mask, uid, gid, &target);
-  if (res == 0) {
-    assert(in == target);
+  int res = _setattr(in, attr, mask, uid, gid);
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Is the removal of target an unrelated cleanup?

File: src/client/Client.cc [hunk 7]
```diff
@@ -9923,20 +9950,17 @@ int Client::ll_create(Inode *parent, const char *name, mode_t mode,
 	        0, 0, 0, NULL, &created, uid, gid);
     if (r < 0)
       goto out;
-
-    if ((!in) && fhp)
-      in = (*fhp)->inode;
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: Why don't we need this logic any more?

**Comment by gregsfortytwo** [comment]
The ref counting this introduces makes a lot of the code paths a lot easier to follow.

But it makes me wonder if we shouldn't just bite the bullet and switch to use of shared_ptrs everywhere — and if not everywhere, why not use them instead of the custom C_InodeRef class?

**Comment by ukernel** [comment]
On Mon, Jun 1, 2015 at 7:24 AM, Gregory Farnum <notifications@github.com>
wrote:

> The ref counting this introduces makes a lot of the code paths a lot
> easier to follow.
>
> But it makes me wonder if we shouldn't just bite the bullet and switch to
> use of shared_ptrs everywhere -- and if not everywhere, why not use them
> instead of the custom C_InodeRef class?
>

we can't use shared_ptr because we need customized release function. we
can instrutive_ptr, but we need to add a new field to class Inode, which
points to class Client.

File: src/client/Client.cc (hunk 1 above)
Comment (ukernel) [inline src/client/Client.cc]: On Mon, Jun 1, 2015 at 6:03 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/client/Client.cc
> <https://github.com/ceph/ceph/pull/4753#discussion_r31394811>:
>
> > +    Inode *tmp = in;
> > +    in = o.in;
> > +    o.in = tmp;
> > +  }
> > +  C_InodeRef& operator=(Inode *in) {
> > +    assert(client);
> > +    C_InodeRef tmp(client, in);
> > +    swap(tmp);
> > +    return *this;
> > +  }
> > +  C_InodeRef& operator=(const C_InodeRef& o) {
> > +    C_InodeRef tmp(o);
> > +    swap(tmp);
> > +    return *this;
> > +  }
> > +  Inode *get() const { return in; }
>
> Why did you want to explicitly invoke a get() function instead of having
> an implicit conversion from C_InodeRef to Inode*?
>
because intrusive_ptr does not have implicit conversion either.

File: src/client/Client.cc (hunk 7 above)
Comment (ukernel) [inline src/client/Client.cc]: On Mon, Jun 1, 2015 at 7:22 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/client/Client.cc
> <https://github.com/ceph/ceph/pull/4753#discussion_r31395431>:
>
> > @@ -9923,20 +9950,17 @@ int Client::ll_create(Inode *parent, const char *name, mode_t mode,
> >  	        0, 0, 0, NULL, &created, uid, gid);
> >      if (r < 0)
> >        goto out;
> > -
> > -    if ((!in) && fhp)
> > -      in = (*fhp)->inode;
>
>
check Client::_create, you can sess that ((!in) && fhp) never be true

File: src/client/Client.cc (hunk 4 above)
Comment (ukernel) [inline src/client/Client.cc]: On Mon, Jun 1, 2015 at 6:07 AM, Gregory Farnum <notifications@github.com>
wrote:

> In src/client/Client.cc
> <https://github.com/ceph/ceph/pull/4753#discussion_r31394841>:
>
> > @@ -5209,32 +5249,32 @@ int Client::get_or_create(Inode *dir, const char* name,
> >    return 0;
> >  }
> >
> > -int Client::path_walk(const filepath& origpath, Inode **final, bool followsym)
> > +int Client::path_walk(const filepath& origpath, C_InodeRef *end, bool followsym)
>
> Why the name change?
>

final is keyword. it looks weired in editor

**Comment by gregsfortytwo** [comment]
> we can't use shared_ptr because we need customized release function. we
can instrutive_ptr, but we need to add a new field to class Inode, which
points to class Client.

Wouldn't the deleter functionality work for this? See for instance the various RemoveOnDelete classes in ceph.git.

File: src/client/Client.cc (hunk 4 above)
Comment (gregsfortytwo) [inline src/client/Client.cc]: Hah, good reason.

**Comment by ukernel** [comment]
providing RemoveOnDelete class for shared_ptr also does not work, because client manipulates inodes's reference count on various cases.

**Comment by gregsfortytwo** [comment]
Mmm, isn't it generally just when linking the inodes to a dentry or filehandle that actually takes a pointer (and could take a shared pointer)?

If not we can always (ugh) use a shared pointer for our internal manipulation that latches on to the manual ref counting we have now for use at interface boundaries.

**Comment by gregsfortytwo** [comment]
@jcsp could you take a look at this and provide your thoughts?

**Comment by jcsp** [comment]
My concerns are broadly the same as greg's, I think:
 * It's a bit odd to pass InodeRefs for some arguments and Inode* for others
 * I would rather avoid the custom smart pointer class

I would suggest using shared_ptr everywhere for the inode pointers, and have Inode carry a reference to Client, so that it can call out to the Client::put_inode stuff from ~Inode.  We are already taking the overhead of a Client* pointer for each inode in the InodeRef class.

Although it's more code to change, I think in the long run it saves effort when trying to explain to people why some inodes are plain pointers and some are a custom pointer class.

**Comment by ukernel** [comment]
On Wed, Jun 10, 2015 at 6:50 PM, John Spray <notifications@github.com>
wrote:

> My concerns are broadly the same as greg's, I think:
>
>    - It's a bit odd to pass InodeRefs for some arguments and Inode* for
>    others
>    - I would rather avoid the custom smart pointer class
>
> I would suggest using shared_ptr everywhere for the inode pointers, and
> have Inode carry a reference to Client, so that it can call out to the
> Client::put_inode stuff from ~Inode. We are already taking the overhead of
> a Client* pointer for each inode in the InodeRef class.
>
> Although it's more code to change, I think in the long run it saves effort
> when trying to explain to people why some inodes are plain pointers and
> some are a custom pointer class.
>

client need to manipulate inodes' reference counts on various cases. (such
as marking caps dirty, adding/dropping the first ll_ref). can shared_ptr
support this?

**Comment by jcsp** [comment]
> client need to manipulate inodes' reference counts on various cases. (such
as marking caps dirty, adding/dropping the first ll_ref). can shared_ptr
support this?

Ah, fair point.

So we could have two types of pointer, a base reference that has the explicit reference counting and then shared_ptrs for our function arguments (which I guess is what greg meant by "If not we can always (ugh) use a shared pointer for our internal manipulation that latches on to the manual ref counting we have now for use at interface boundaries.").  However, I'm not sure how valuable that is compared with just having the one (custom) smart pointer type.

**Comment by ukernel** [comment]
there are two set of libcephfs functions. One set only receive path or fd as parameter and do not receive/return inode pointer. Another set (which starts with ll_) may return inode pointer and they increase inode's ll_ref when inode is returned. users of ll_xxx functions manage reference counting themself (for the fuse case, reference counting is managed by kernel. shared pointer can't help) .

**Comment by gregsfortytwo** [comment]
Right, but those ll* functions are an interface boundary we can work with. We can internally use a shared pointer everywhere, and only turn it into a regular pointer at those boundaries.
We still need to have a manual refcount on the inode, but instead of tweaking it manually in many places, we tweak it only in the ll* functions and when creating a new shared pointer (which we can add to a registry to make sure we don't have dueling shared pointers). And the shared pointer deleter can simply decrement the refcount.

Or, d'oh, we can make use of the intrusive_ptr which is designed for classes wit manual recounts. http://www.boost.org/doc/libs/1_57_0/libs/smart_ptr/intrusive_ptr.html

**Comment by gregsfortytwo** [comment]
@jcsp, please take a look.

**Comment by jcsp** [comment]
I'm happy with this -- looks like exactly what intrusive_ptr is designed for.  If there are any places we can replace more Inode* with InodeRef then we can do that in subsequent patches.

Ready to merge this once it's rebased on master -- I merged wip-12088 first because that one might need backporting.

**Comment by ukernel** [comment]
ping @jcsp
